content/en/blog/_posts/2024-09-17-devsecops-in-action-k8s-secrets/index.md

```diff
@@ -0,0 +1,554 @@
+			return "", false
+		}
+	}
+	return "", false
+}
+```
+
+Magic happens in the `httpHandle` function. Environment variable value is passed
+into function, if it's a file it will be read and current environment variable
+value will be overridden by a new value that was read from the file.
+
+[`cmd/main.go`](https://github.com/efrikin/devsecops-in-action-kubernetes-secrets/blob/8c8177b4955ce9cd9686c2746e5e2a810b9c2f62/cmd/main.go#L45-L59)
+```go
+func httpHandle(w http.ResponseWriter, r *http.Request) {
+	if !changed {
+		if envVal, isExistEnv = os.LookupEnv(envVar); isExistEnv {
+			if filePath, isExistFile = fileExist(envVal); isExistFile {
+				body, err := os.ReadFile(filePath)
+				if err != nil {
+					log.Fatalf("unable to read file: %v", err)
+				}
+				os.Setenv(envVar, string(body))
+				envVal = os.Getenv(envVar)
+				changed = true
+			}
+		}
+	}
+}
+```
+
+Finally, if define `DEMO_SECRET__PASSWD=file:///secrets/secret.file`, deploy
+application and make request to `root` location, response will contain file
+content which is defined in the environment variable:
+
+```bash
+curl -v localhost:8080
+```
+
+The output is similar to this:
+
+```text
+...
+> GET / HTTP/1.1
+> Host: localhost:8080
+> User-Agent: curl/8.5.0
+> Accept: */*
+>
+< HTTP/1.1 200 OK
+...
+PVHBD5I21osTOX9i
+```
+
+Let's check process environment:
+
+```bash
+# strings /proc/1/environ
+```
+
+The output is similar to this:
+
+```text
+...
+DEMO_SECRET__PASSWD=file:///secrets/secret.file
+...
+```
+
+Excellent! Environment variable still contains the value which was defined
+during the deployment.
+
+Let's check the secret file:
+
+```bash
+# cat /secrets/secret.file
+
+#
+```
+
+The file is empty! After readiness probe passed, the file was truncated.
+
+If define `DEMO_SECRET__PASSWD=/secrets/secret.file` (without prefix) response
+will contain the file name which was defined in the environment variable:
+
+```bash
+# curl -v localhost:8080
+```
+
+The output is similar to this:
+
+```text
+...
+> GET / HTTP/1.1
+> Host: localhost:8080
+> User-Agent: curl/8.5.0
+> Accept: */*
+>
+< HTTP/1.1 200 OK
+...
+/secrets/secret.file
+```
+
+```bash
+# strings /proc/1/environ
+```
+
+```text
+...
+DEMO_SECRET__PASSWD=/secrets/secret.file
+...
+```
+
+```bash
+# cat /secrets/secret.file
+```
+
+```text
+cat: /secrets/secret.file: No such file or directory
+```
+
+`/secrets/secret.file` is a string interpreted as sensitive data, that's why
+the file wasn't injected.
+
+The attacker's life has become more complicated because now in order to
+access sensitive data it's necessary either to make a process dump or to use a
+debugger tool. In order to show debug process, first it's necessary to make
+build with disabled compiler optimization flags and run `gdb`:
+
+```bash
+go build -gcflags=all="-N -l" cmd/stage-03/main.go
+echo 'MySecret' > secret.file
+DEMO_SECRET__PASSWD=file://secret.file gdb main
+```
+
+Let's check current process environment:
+
+```bash
```

Comment:
This isn't shell; please drop the `bash` suffix.

Reply:
Fixed